### PR TITLE
[ci] use Ruff linter instead of isort

### PR DESCRIPTION
**Pull request by StrikerRUS**
Let Ruff work with imports sorting directly.
This change speeds up the pre-commit job, avoids situations where [ruff and isort are fighting each other](https://github.com/microsoft/LightGBM/pull/6317#discussion_r1491925349) and allows to consistenly configure only one tool.

Refer to: https://docs.astral.sh/ruff/faq/#how-does-ruffs-import-sorting-compare-to-isort.

Also, bump versions of pre-commit libraries.

### CODE REVIEW

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -2507,13 +2507,13 @@ def _compare_params_for_warning(
         compare_result : bool
           Returns whether two dictionaries with params are equal.
         """
-        for k in other_params:
+        for k, v in other_params.items():
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:2510]: Fix [PLC0206](https://docs.astral.sh/ruff/rules/dict-index-missing-items/#dict-index-missing-items-plc0206) in new Ruff version.

File: python-package/lightgbm/basic.py [hunk 2]
```diff
@@ -2507,13 +2507,13 @@ def _compare_params_for_warning(
...
             if k not in ignore_keys:
-                if k not in params or params[k] != other_params[k]:
+                if k not in params or params[k] != v:
                     return False
-        for k in params:
+        for k, v in params.items():
```
Comment (StrikerRUS) [inline python-package/lightgbm/basic.py:2514]: Fix [PLC0206](https://docs.astral.sh/ruff/rules/dict-index-missing-items/#dict-index-missing-items-plc0206) in new Ruff version.

File: python-package/pyproject.toml [hunk 3]
```diff
@@ -154,6 +143,7 @@ select = [
     "D",
     # pycodestyle
     "E",
+    "W",
```
Comment (StrikerRUS) [inline python-package/pyproject.toml:146]: Add [pycodestyle Warnings](https://docs.astral.sh/ruff/rules/#warning-w)

File: python-package/pyproject.toml [hunk 4]
```diff
@@ -166,11 +156,13 @@ select = [
     "SIM401",
     # flake8-print
     "T",
+    # isort
+    "I",
```
Comment (StrikerRUS) [inline python-package/pyproject.toml:160]: Enable [isort](https://docs.astral.sh/ruff/rules/#isort-i).

File: python-package/pyproject.toml [hunk 5]
```diff
@@ -196,3 +188,6 @@ select = [
 
 [tool.ruff.lint.pydocstyle]
 convention = "numpy"
+
+[tool.ruff.lint.isort]
+known-first-party = ["lightgbm"]
```
Comment (StrikerRUS) [inline python-package/pyproject.toml:193]: https://docs.astral.sh/ruff/faq/#how-does-ruff-determine-which-of-my-imports-are-first-party-third-party-etc

File: python-package/lightgbm/basic.py (hunk 2 above)
Comment (jameslamb) [reply python-package/lightgbm/basic.py:2514]: oh wow, I'm impressed that a linter could look in this far and find something like this! Really nice.

**Review by jameslamb — APPROVED** [review]
Really like this, thanks very much1

File: python-package/pyproject.toml (hunk 3 above)
Comment (jameslamb) [reply python-package/pyproject.toml:146]:
```suggestion
    # pycodestyle (warnings)
    "W",
```

Sorry, missed this before I approved.

Could you:

* add a comment like this one
* change the comment over `"E"` `# pycodestyle (errors)`
* keep this list alphabetically organized by rule code (so move this `"W"` to the end, and put `isort` after `"F"`

I think that makes it a bit easier to find the relevant entry in the config and map it to the type of check. 

Very very minor suggestion, so if you disagree at all then just merge this.

Comment (StrikerRUS) [reply python-package/pyproject.toml:146]: Sure, will do.

Comment (StrikerRUS) [reply python-package/pyproject.toml]: Done in 4cb519e76857a0a2f261948034d0ab098e63129b. Is it OK now?

**Review by jameslamb — APPROVED** [review]
re-approving

**Pull request merged**